pkg/providers/cloudstack/cloudstack.go

```diff
@@ -0,0 +1,765 @@
+		controlPlaneMachineSpec = &machineConfigs[clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec
+	}
+	if len(clusterConfig.Spec.WorkerNodeGroupConfigurations) > 0 && clusterConfig.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef != nil && machineConfigs[clusterConfig.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name] != nil {
+		workerNodeGroupMachineSpec = &machineConfigs[clusterConfig.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name].Spec
+	}
+	if clusterConfig.Spec.ExternalEtcdConfiguration != nil {
+		if clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef != nil && machineConfigs[clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name] != nil {
+			etcdMachineSpec = &machineConfigs[clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec
+		}
+	}
+	return &cloudstackProvider{
+		datacenterConfig:      datacenterConfig,
+		machineConfigs:        machineConfigs,
+		clusterConfig:         clusterConfig,
+		providerKubectlClient: providerKubectlClient,
+		writer:                writer,
+		selfSigned:            false,
+		templateBuilder: &CloudStackTemplateBuilder{
+			datacenterConfigSpec:       &datacenterConfig.Spec,
+			controlPlaneMachineSpec:    controlPlaneMachineSpec,
+			workerNodeGroupMachineSpec: workerNodeGroupMachineSpec,
+			etcdMachineSpec:            etcdMachineSpec,
+			now:                        now,
+		},
+		skipIpCheck: skipIpCheck,
+	}
+}
+
+func (p *cloudstackProvider) UpdateKubeConfig(_ *[]byte, _ string) error {
+	// customize generated kube config
+	return nil
+}
+
+func (p *cloudstackProvider) BootstrapClusterOpts() ([]bootstrapper.BootstrapClusterOption, error) {
+	env := map[string]string{}
+	execConfig, err := decoder.ParseCloudStackSecret()
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse environment variable exec config: %v", err)
+	}
+	if p.clusterConfig.Spec.ProxyConfiguration != nil {
+		noProxy := fmt.Sprintf("%s,%s", p.clusterConfig.Spec.ControlPlaneConfiguration.Endpoint.Host, execConfig.CloudStackManagementUrl)
+		for _, s := range p.clusterConfig.Spec.ProxyConfiguration.NoProxy {
+			if s != "" {
+				noProxy += "," + s
+			}
+		}
+		env["HTTP_PROXY"] = p.clusterConfig.Spec.ProxyConfiguration.HttpProxy
+		env["HTTPS_PROXY"] = p.clusterConfig.Spec.ProxyConfiguration.HttpsProxy
+		env["NO_PROXY"] = noProxy
+	}
+	return []bootstrapper.BootstrapClusterOption{bootstrapper.WithEnv(env)}, nil
+}
+
+func (p *cloudstackProvider) Name() string {
+	return constants.CloudStackProviderName
+}
+
+func (p *cloudstackProvider) DatacenterResourceType() string {
+	return eksaCloudStackDeploymentResourceType
+}
+
+func (p *cloudstackProvider) MachineResourceType() string {
+	return eksaCloudStackMachineResourceType
+}
+
+func (p *cloudstackProvider) setupSSHAuthKeysForCreate() error {
```

Comment:
ssh keys work the same way for cloudstack as they do for vmware right? Is there some wireguard like setup required to ssh?

Reply:
Oh so how do you log into the node to run commands? I was assuming you can use ssh to log in just like we do for vsphere.

Also, this looks like the same as vsphere, so maybe we can move this to common too.

Reply:
You would ssh using the user and corresponding private key to the ones passed in in the cluster spec. I can take a swing at extracting out this duplicate logic. It's a little messy since it's not read-only but maybe the common function can return a list of ssh keys and we'll set them in the invoking function